src/Chapter4.hs

```diff
@@ -489,6 +503,41 @@ Implement the 'Applicative' instance for our 'List' type.
   type.
 -}
 
+instance Applicative List where
+  pure a = Cons a Empty
+  (<*>) :: List (a -> b) -> List a -> List b
+  Empty <*> _ = Empty
+  Cons f xfs <*> ys = lappend first second
+    where first = fmap f ys
+          second = xfs <*> ys
+
+{-
+-- Examples
+>>> c1 = Cons 1 (Cons 2 Empty)
+>>> f1 = Cons (+1) (Cons (+2) (Cons ((+3)) Empty))
+>>> f1 <*> c1
+Cons 2 (Cons 3 (Cons 3 (Cons 4 (Cons 4 (Cons 5 Empty)))))
+-}
+
+lappend :: List a -> List a -> List a
+lappend Empty Empty = Empty
+lappend xs Empty = xs
+lappend Empty xs = xs
+lappend (Cons x Empty) (Cons y Empty) = Cons x (Cons y Empty)
+lappend (Cons x xs) (Cons y Empty) = Cons x (radd xs y)
+lappend (Cons x xs) (Cons y ys) = ladd x (lappend xs (ladd y ys))
+
+ladd :: a -> List a -> List a
+ladd x Empty = Cons x Empty
+ladd x (Cons y Empty) = Cons x (Cons y Empty)
+ladd x (Cons y ys) = Cons x (Cons y ys)
+
+radd :: List a -> a -> List a
+radd Empty x = Cons x Empty
+radd (Cons y Empty) x = Cons x (Cons y Empty)
+radd (Cons y ys) x  = Cons x (Cons y ys)
```

Comment:
Isn't this function is the same as `ladd` but with the arguments flipped?
You can specify it like this:

```haskell
radd l x = ladd x l
```

or use `flip`:

```haskell
radd = flip ladd
```

Reply:
👍